src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.cs

```diff
@@ -307,14 +307,57 @@ private void AddClientConnection(ClientConnectionContext connection, OpenConnect
         }
 
         private async Task ProcessIncomingMessageAsync(ClientConnectionContext connection)
+        {
+            Exception exception = null;
+            try
+            {
+                await connection.ApplicationTask;
+            }
+            catch (Exception e)
+            {
+                exception = e;
+            }
+            finally
+            {
+                // If we aren't already aborted, we send the abort message to the service
+                if (connection.AbortOnClose)
+                {
+                    // Inform the Service that we will remove the client because SignalR told us it is disconnected.
+                    var serviceMessage = new CloseConnectionMessage(connection.ConnectionId, errorMessage: exception?.Message);
+                    await WriteAsync(serviceMessage);
+                    Log.CloseConnection(Logger, connection.ConnectionId);
+                }
+            }
+        }
+
+        private async Task ProcessApplicationTaskAsync(ClientConnectionContext connection)
+        {
+            // Wait for the application task to complete
+            // application task can end when exception, or Context.Abort() from hub
+            var app = ProcessApplicationTaskAsyncCore(connection);
+
+            var cancelTask = connection.ApplicationAborted.AsTask();
+            var task = await Task.WhenAny(app, cancelTask);
+
+            if (task != app)
+            {
+                // cancel the application task, to end the outgoing task
+                connection.Application.Input.CancelPendingRead();
+                Log.ApplicationTaskCancelled(Logger);
+            }
+
+            await task;

```

Comment:
Why `await task` again? Should we wait for `app` instead?

Reply:
await task for the task to end or throw